ReSharper.FSharp/src/FSharp.Psi.Services/src/Util/SignatureFileUtil.fs

```diff
@@ -0,0 +1,172 @@
+        OpenStatements: IOpenStatement list
+        SigModule: INamedModuleDeclaration
+    }
+    
+    member this.SigFile = this.SigModule.Parent :?> IFSharpSigFile
+
+let tryMkBindingSignature
+    (letBindings: ILetBindings)
+    (moduleDecl: IModuleDeclaration)
+    : SignatureBindingResponse option
+    =
+    if isNull moduleDecl then None else
+    
+    let allDecls = moduleDecl.DeclaredElement.GetDeclarations() |> Seq.cast<INamedModuleDeclaration> |> Seq.toArray
+    if allDecls.Length <> 2 then None else
+
+    let implDecl = Array.find (fun (nmd: INamedModuleDeclaration) ->
+        match nmd.Parent with
+        | :? IFSharpImplFile -> true
+        | _ -> false) allDecls
+
+    let sigDecl = Array.find (fun (nmd: INamedModuleDeclaration) ->
+        match nmd.Parent with
+        | :? IFSharpSigFile -> true
+        | _ -> false) allDecls
+
+    let elementFactory = sigDecl.CreateElementFactory()
+    let binding = letBindings.Bindings.First()
+    let refPat = binding.HeadPattern.As<IReferencePat>()
+    let name = refPat.ReferenceName.Identifier.Name
+    let symbolUse = refPat.GetFcsSymbolUse()
+    if isNull symbolUse then None else
+
+    let mfv = symbolUse.Symbol :?> FSharpMemberOrFunctionOrValue
+    let types = FcsTypeUtil.getFunctionTypeArgs true mfv.FullType
+    let parameters = binding.ParametersDeclarations
+
+    let rec getTypeName (t:FSharpType) : string =
+        if t.HasTypeDefinition then
+            if Seq.isEmpty t.GenericArguments then
+                t.TypeDefinition.DisplayName
+            else
+                let isPostFix =
+                    set [| "list"; "option"; "array" |]
+                    |> Set.contains t.TypeDefinition.DisplayName
+                    
+                if isPostFix then
+                    let ga = t.GenericArguments[0].GenericParameter
+                    let tick = if ga.IsSolveAtCompileTime then "^" else "'"
+                    $"{tick}{ga.DisplayName} {t.TypeDefinition.DisplayName}"
+                else
+                    let args = Seq.map getTypeName t.GenericArguments |> String.concat ","
+                    $"{t.TypeDefinition.DisplayName}<{args}>"
+        elif t.IsGenericParameter then
+            let tick = if t.GenericParameter.IsSolveAtCompileTime then "^" else "'"
+            $"{tick}{t.GenericParameter.DisplayName}"
+        elif t.IsFunctionType then
+            let rec visit (t: FSharpType) : string seq =
+                if  t.IsFunctionType then
+                    Seq.collect visit t.GenericArguments
+                else
+                    Seq.singleton (getTypeName t)
+            
+            visit t
+            |> String.concat " -> "
+            |> sprintf "(%s)"
+        elif t.IsTupleType then
+            t.GenericArguments
+            |> Seq.map getTypeName
+            |> String.concat " * "
+        else
+            "???"
+
+    let rec visitAccessPath (t: FSharpType) : string list =
```

Comment:
There are probably better ways to do all this.
I'm all ears if you can give some pointers 😸.

Reply:
`RequireQualifiedAccess` and `AutoOpen` attributes may make us want insert a different namespace/module open than the one that a type is defined in. `OpensUtil.getModuleToOpen` takes these things into account.